### ISSUE TITLE
SocketAddress with arbitrary AddressFamily throws PNSE on Unix

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#37997]
This succeeds on Windows:
```C#
new SocketAddress((AddressFamily)12345).ToString()
```
returning:
```
"12345:32:{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}"
```
but on Linux and macOS it throws:
```
Unhandled Exception of Type System.PlatformNotSupportedException
Message :
System.PlatformNotSupportedException : Operation is not supported on this platform.
Stack Trace :
   at System.Net.SocketAddressPal.ThrowOnFailure(Error err) in /_/src/Common/src/System/Net/SocketAddressPal.Unix.cs:line 51
   at System.Net.SocketAddressPal.SetAddressFamily(Byte[] buffer, AddressFamily family) in /_/src/Common/src/System/Net/SocketAddressPal.Unix.cs:line 76
   at System.Net.SocketAddress..ctor(AddressFamily family, Int32 size) in /_/src/Common/src/System/Net/SocketAddress.cs:line 96
```

**Comment by wfurt** [comment]
How come it works on Windows? Did OS created socket with unknown family? When I tested Netlink and CanBus I would get error.

**Comment by stephentoub** [comment]
> How come it works on Windows? Did OS created socket with unknown family? When I tested Netlink and CanBus I would get error.

On Windows it doesn't create a Socket or go to the OS at all: it just fills bits into the buffer:
https://github.com/dotnet/corefx/blob/e3a4edabd12b5f7dc9be20136cc867bbe67f1f7c/src/Common/src/System/Net/SocketAddressPal.Windows.cs#L19-L28

In contrast, on Unix it does:
https://github.com/dotnet/corefx/blob/e3a4edabd12b5f7dc9be20136cc867bbe67f1f7c/src/Common/src/System/Net/SocketAddressPal.Unix.cs#L73
which calls:
https://github.com/dotnet/corefx/blob/6f6790fa8219359f2acbe4b0624a3fa9a0a6943b/src/Native/Unix/System.Native/pal_networking.c#L583
which will return EAFNOSUPPORT:
https://github.com/dotnet/corefx/blob/6f6790fa8219359f2acbe4b0624a3fa9a0a6943b/src/Native/Unix/System.Native/pal_networking.c#L585
and result in the PlatformNotSupportedException.

**Comment by wfurt** [comment]
What do you suggest? Defer creating OS socket? 
Socket with unknown family would not be useful for any real usage, right?

**Comment by tmds** [comment]
> On Windows it doesn't create a Socket or go to the OS at all: it just fills bits into the buffer:

Do the enum values of `AddressFamily` match the native AF_xxx values on Windows?

**Comment by tmds** [comment]
> Do the enum values of AddressFamily match the native AF_xxx values on Windows?

Yes, this seems to be the case. That's why it makes sense to fill in the address family in the Windows implementation. On Linux, you can only fill in what you can map from 'Windows values', PNSE makes sense then.

The recently added values don't map directly with Windows AF_xxx values:

https://github.com/dotnet/corefx/blob/e3a4edabd12b5f7dc9be20136cc867bbe67f1f7c/src/System.Net.Primitives/src/System/Net/Sockets/AddressFamily.cs#L46-L48

Maybe we should start these at a higher offset, e.g. 65536 (so they take more than 2 bytes).
SocketAddress on Windows can throw PNSE for these.

**Comment by tmds** [comment]
@stephentoub @wfurt based on my earlier comment, we may want to change the values that added in `AddressFamily`. If such a change is wanted, it should probably be for 3.0?

**Comment by stephentoub** [comment]
Thanks, @tmds.  @karelz, @wfurt, what do you want to do with this?

**Comment by davidsh** [comment]
>Thanks, @tmds. @karelz, @wfurt, what do you want to do with this?

These recently added APIs were done for .NET Core 3.0. We need to fix this before we ship .NET Core 3.0.

**Comment by wfurt** [comment]
I don't think the addition matters that much @davidsh. You can have yet another protocol number and that will behave different way on Unix and Windows. (like the example with 12345) I'm not aware of any practical impact so this will probably not meet 3.0 bar.

**Comment by tmds** [comment]
It's not major, this is the practical impact:
- `AddressFamily` values can no longer match with Windows `AF_xxx` values for values `30` and higher. So if such values are added, they will need some mapping (like we do on non-Windows).
- `SocketAddress` should throw on PNSE on Windows for these new values, which it doesn't.

**Comment by karelz** [comment]
I think we should try to address it and run it by shiproom.

**Comment by davidsh** [comment]
@wfurt @tmds @stephentoub Can one of you please submit a PR with proposed fixes for this (i.e. different enum values, etc.)?  We'd like something concrete to discuss with shiproom.

**Comment by wfurt** [comment]
1) I think key decision is if we want to support SocketAddress of arbitrary protocols e.g. `SocketAddress((AddressFamily)12345)`. That can make life easier for anybody looking for early support of new protocol and matches Windows and Framework behavior (tested on 4.7). 
If we say yes here, then we assume users know what they doing and we can IMHO simply marshal given value to sa_family. That seems common to all OSes. If we decide we do not to support arbitrary AF, we throw on Windows to match Unix. As @tmds suggested we would also throw on Windows for three new families as there is no mapping to OS values. 

2) I don't think we should change `AddressFamily` it self. That is public enum and it seems like we would start encoding some low-level OS knowledge to it's values.  For 3.0 on Windows we should either throw or pass raw values like we do for 12345.

3) If Windows support new protocols in future we should add mapping function to PAL like we do on Unix. That can still be pretty simple but it would give us proper abstraction. 

4) As @stephentoub pointed out `SocketAddress` does not depend on any OS functions.
That essentially defers failure to point when socket of given AF is created. I tried to run my test code with Raw socket on Windows and I get `An address incompatible with the requested protocol was used`. If we want to, we can explicitly throw PNSP on Windows and OSX for new address families in `Socket` constructor. On Linux we need to give it try as the support depends on loaded kernel modules.

Please let me know what is your preferred approach for four items above @davidsh @tmds @stephentoub and I'll work on PR once we reach consensus.

**Comment by karelz** [comment]
@davidsh @stephentoub can you please help decide which option is best for 3.0 vs. Future?
Thanks!

**Comment by tmds** [comment]
> If we say yes here, then we assume users know what they doing and we can IMHO simply marshal given value to sa_family.

I would prefer this as a separate API that makes explicit the passed value is the raw value, e.g. accept an `int` and not an `AddressFamily`.

> That is public enum and it seems like we would start encoding some low-level OS knowledge to it's values.

That has been implicitly the case so far. The values match Windows AF_ constants.

> SocketAddress does not depend on any OS functions.

`SocketAddress` is the OS specific representation of an `EndPoint`.

**Comment by karelz** [comment]
@stephentoub can you please help us decide on the next steps?
@scalablecory should have time to do the work ...

Thanks!

**Comment by stephentoub** [comment]
The pre-3.0 values all map directly to Windows, but need to be translated for Unix.

The new values added in 3.0 don't have direct Windows equivalents, but if they were ever added to Windows in the future, there's no guarantee they would have the same OS values.  Plus in theory Windows could add additional values in the future that end up conflicting with these.

All of that points to needing to add an explicit mapping for the new values on Windows, i.e. anything that's been there for a while can be passed straight through, but the new values need to be mapped, or if there is no mapping, throw.

I don't think it matters whether we change the actual enum values.  We simply add a check on Windows where these values are used.  Any value <= max we allow it to be passed through.  Any value > max we treat as needing mapping/throwing.

**Comment by wfurt** [comment]
I like the proposal. It seems like value 32 is used for AF_BTH (Bluetooth) on Windows and that conflicts with AF_CAN. 

https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-socket

I will add simple mapping function as @stephentoub suggested. With that, we do not need to worry about keeping everything in sync.

**Comment by karelz** [comment]
So, we kept the issue in 3.0 milestone, but the PR went only into master (5.0). Reopening.
Do we need to port it into release/3.0, or is 5.0-only fix ok?

**Comment by stephentoub** [comment]
> is 5.0-only fix ok?

No. If it doesn't get into 3.0, it needs to be reverted from master. It's a breaking change.

@danmosemsft, this is the kind of thing I was mentioning.

**Comment by danmosemsft** [comment]
Despite reading all this carefully I'm missing something. 
 `new SocketAddress((AddressFamily)12345).ToString();` succeeds on 2.2. So if we want to break it, it doesn't have to break in 3.0. It could be in 5.0. Right?

**Comment by stephentoub** [comment]
@danmosemsft, while true, there are new named enum values in the range that should fail but don't. It doesn't require you to cast arbitrary values.

**Comment by danmosemsft** [comment]
ah, of course. Well, I'm willing to take it to tactics if there's a PR + template on Tuesday. I could see it getting rejected at this point though.

**Comment by wfurt** [comment]
note that `new SocketAddress((AddressFamily)32)` can produce valid Bluetooth socket on certain versions of Windows with 2.x. Windows can go up to value 34 (AF_MAX 35) (from current ws23def.h) 

Only safe way how to avoid conflict while preserving existing behavior is to avoid  conflict. While #39785 made it more consistent, it can possibly break users with more validation. 

We can either skip values currently used on Windows or use fact that sa_family is 16 bits and use ushort.MaxValue + 1.

With that we can throw when new enum values are used on WIndows while preserving ability to cast int to values supported by OS.

**Comment by tmds** [comment]
> We can either skip values currently used on Windows or use fact that sa_family is 16 bits and use ushort.MaxValue + 1.

This matches my suggestion https://github.com/dotnet/corefx/issues/37997#issuecomment-497063823.

**Comment by wfurt** [comment]
approved by Tactics. 
closed by #40235.
